lib/net.js

```diff
@@ -667,6 +640,16 @@ function onread(nread, buffer) {
     maybeDestroy(self);
```

Comment:
I wonder if this `if` statement can be removed. `readable` is automatically set to `false` when `'end'` is emitted and `maybeDestroy()` is called by the handler below.

Reply:
Yeah, but `end` hasn’t necessarily been emitted here. But you’re right, the handler below was supposed to be in the `else` block for this.

Reply:
Yes but if I'm not wrong `'end'` will not be emitted on the same tick, so why aren't we only relying on `'end'` event? Is it to avoid attaching a listener?

Reply:
> Is it to avoid attaching a listener?

Yes, I assume that’s it. Maybe it isn’t that important?

Reply:
I don't know, I don't think it will make a big difference but there is also no real advantage in removing the "fast case" so let's keep it as is :)

Reply:
FWIW we are now emitting [`'_socketEnd'`](https://github.com/nodejs/node/blob/af5632e4f40b2462139ae1534bc5ef99ce39ebbb/lib/net.js#L673) and [attaching the listener](https://github.com/nodejs/node/blob/af5632e4f40b2462139ae1534bc5ef99ce39ebbb/lib/net.js#L368-L374) all the time as `'end'` is not emitted on the same tick.

I also _think_ this [branch](https://github.com/nodejs/node/blob/af5632e4f40b2462139ae1534bc5ef99ce39ebbb/lib/net.js#L665-L668) can create a race where `'close'` can be emitted before `'end'` as [`_handle.close()`](https://github.com/nodejs/node/blob/af5632e4f40b2462139ae1534bc5ef99ce39ebbb/lib/net.js#L592-L595) is called before the `'end'` event is emitted if [`maybeDestroy()`](https://github.com/nodejs/node/blob/af5632e4f40b2462139ae1534bc5ef99ce39ebbb/lib/net.js#L667) succeeds.

TL;DR I think it's better to always rely on the `'end'` event.

Reply:
> TL;DR I think it's better to always rely on the `'end'` event.

I am cool with that if we can agree that the fast-pathing is not worth it … I like the simplicity :)

Updated!